prism_compile.c

```diff
@@ -970,16 +1022,21 @@ pm_setup_args(pm_arguments_node_t *arguments_node, int *flags, struct rb_callinf
                       break;
                   }
                   else {
-                      *kw_arg = rb_xmalloc_mul_add(len, sizeof(VALUE), sizeof(struct rb_callinfo_kwarg));
-                      *flags = VM_CALL_KWARG;
-                      (*kw_arg)->keyword_len = (int) len;
+                    int compiled = pm_compile_keyword_arg(iseq, ret, keyword_arg, kw_arg, flags, src, scope_node, dummy_line_node, parser);
```

Comment:
The way I'm understanding this code is that we're doing two passes over the `keyword_arg->elements` list. In the first pass (currently done within `pm_compile_keyword_arg` L882-894) we're looking to determine whether it contains at least one non-static literal value. This is the true / false that we're returning. In the true case, we do the compilation and flag setting in `pm_compile_keyword_arg`. In the false case, we do the compilation and flag setting here.

I think we could clean up this code a little bit by making the two passes more explicit, and probably inlining them both here (or at least being consistent about what happens where).

Pass one:
- We set a bool `arguments_contain_all_keywords` or a better name. We don't do any flag setting or compilation.

Pass two:
- We have an if / else based on the value of `arguments_contain_all_keywords`
- In each case we set flags appropriately
- In each case we loop over the arguments again, and compile according to the specific case
- We set  values on `kw_arg` appropriately

I think this will make the code more readable / understandable. It _also_ will have the added benefit of allowing us to pull the first pass into Prism itself. We should know that information at parse time, and can set it as a flag on the arguments node.

Reply:
We already set `PM_NODE_FLAG_STATIC_LITERAL` in the parser for this reason. I think you should be able to just use that directly instead of the first pass

Reply:
Wait, hold on, I did know about the `PM_NODE_FLAG_STATIC_LITERAL` since that's exactly what I am using to see if the keys of the `AssocNode`s inside the `KeywordHashNode` are static literals or not.

@kddnewton Are you saying that the `KeywordHashNode` would also have a `PM_NODE_FLAG_STATIC_LITERAL` flag set if all of the elements inside have static literal keys?

Reply:
It should, yeah. Unless I'm misremembering. I know we did it for arrays and hashes, but maybe we missed the keyword hash node?

Reply:
Yeah, I don't think a KeywordHashNode has that flag. But also, it doesn't seem like that's exactly the logic of this? It seems like we're checking that each node is an AssocNode _and_ a static literal. 

I guess it depends how we'd implement static_literal on KeywordHashNode. But I'd think something like this `method_call(**{})` would have a static literal KeywordHashNode, but not fit into the case that the code describes

Reply:
I don't think that would be a static literal, because that would contain an assoc-splat. (In that case the inner hash _would_ be a static literal.)

Reply:
Ok, let me summarize the way forward that I'll take:
1. I'll inline the logic into the original method and simplify as @jemmaissroff suggested and try to ship this PR.
2. I'll make another PR on `prism` to add the `PM_NODE_FLAG_STATIC_LITERAL` flag to `KeywordHashNode`s if all elements are `AssocNode`s with static literal keys (we don't care if the values are static literals or not).
3. Once that work is done, I'll come back and make another PR on `ruby/ruby` to use that flag for this logic.

Does that sound good to all of you?

Reply:
Made this refactor

Reply:
Actually @paracycle I think we should skip the flag on keyword hash node. I think you should do it as you have here.

STATIC_LITERAL is enabled on a couple of nodes:

* false
* float
* imaginary
* rational
* integer
* nil
* regexp
* source encoding
* source file
* source line
* string (if frozen)
* symbol
* true

I think all of these intuitively make sense. Then we have the containers. It gets a little weird:

* arrays - if they contain all static literals AND do not contain any other containers
* assocs - if the key and value are static literals AND do not contain any other containers
* hash - if all elements are both static literals AND do not contain any other containers
* range - if both left and right are either omitted, nil, or integers

I think the range one is already weird, but at least somewhat consistent. I think adding it for keyword hash where it would only check the keys would be very surprising. 

That being said, I'm very okay with adding a _different_ flag from static literal to keyword hashes that indicates that all of its keys are static literal. I just don't want to reuse that same flag.

Also I kind of want to rename the range flag for the same reason.

Reply:
@kddnewton Thanks. I was thinking about the same thing regarding the `STATIC_LITERAL` flag on keyword hash nodes representing all assoc nodes with static literal keys only and it felt weird to me as well.

I will define a different flag (something like `PM_NODE_FLAG_STATIC_LITERAL_ASSOC_NODE_KEYS` or similar -- open to suggestions) to denote this on keyword hashes, and I'll add/remove the flag inside `pm_keyword_hash_node_elements_append` depending on the element being added.

Does that sound good?